File: create/hub/module_sanitizer.go
package hub

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

var SectionOrder = []string{
	"General",
	"Host",
	"Rule",
	"URL Rewrite",
	"Map Local",
	"Script",
	"Body Rewrite",
	"Header Rewrite",
	"MITM",
}

var HeaderKeysOrder = []string{
	"name",
	"desc",
	"author",
	"icon",
	"category",
	"tag",
	"date",
	"arguments",
	"arguments-desc",
	"system-proxy",
	"ability",
	"update-interval",
}

var (
	metaLineRe      = regexp.MustCompile(`(?i)^#!\s*([A-Za-z0-9_-]+)\s*[=:]\s*(.*)$`)
	scriptNameRe    = regexp.MustCompile(`(?i)\bname\s*=\s*([^,\s]+)`)
	scriptLabelRe   = regexp.MustCompile(`(?i)^(.+?)\s*=\s*type=`)
	scriptPathRe    = regexp.MustCompile(`(?i)\bscript-path\s*=\s*([^,\s]+)`)
	scriptPatternRe = regexp.MustCompile(`(?i)\bpattern\s*=\s*([^,]+)`)
)

type ModuleSection struct {
	Name  string
	Lines []string
}

func ParseModule(text string) (map[string]string, []ModuleSection) {
	headerMeta := make(map[string]string)
	var sections []ModuleSection
	currentName := ""
	var currentLines []string
	inBody := false

	lines := strings.Split(text, "\n")
	for _, raw := range lines {
		stripped := strings.TrimSpace(raw)
		if strings.HasPrefix(stripped, "[") && strings.HasSuffix(stripped, "]") && !strings.HasPrefix(stripped, "#") {
			if currentName != "" {
				sections = append(sections, ModuleSection{Name: currentName, Lines: currentLines})
			}
			currentName = stripped[1 : len(stripped)-1]
			currentLines = []string{}
			inBody = true
			continue
		}

		if !inBody {
			match := metaLineRe.FindStringSubmatch(stripped)
			if match != nil {
				headerMeta[strings.ToLower(match[1])] = strings.TrimSpace(match[2])
			}
			continue
		}

		if currentName != "" {
			currentLines = append(currentLines, strings.TrimRight(raw, "\r"))
		}
	}

	if currentName != "" {
		sections = append(sections, ModuleSection{Name: currentName, Lines: currentLines})
	}
	return headerMeta, sections
}

func dedupeKey(section string, line string) string {
	stripped := strings.TrimSpace(line)
	if stripped == "" || strings.HasPrefix(stripped, "#") {
		return ""
	}

	if section == "Script" {
		labelMatch := scriptLabelRe.FindStringSubmatch(stripped)
		if labelMatch != nil {
			return fmt.Sprintf("script:label:%s", strings.TrimSpace(labelMatch[1]))
		}
		nameMatch := scriptNameRe.FindStringSubmatch(stripped)
		if nameMatch != nil {
			return fmt.Sprintf("script:name:%s", nameMatch[1])
		}
		pathMatch := scriptPathRe.FindStringSubmatch(stripped)
		patternMatch := scriptPatternRe.FindStringSubmatch(stripped)
		if pathMatch != nil && patternMatch != nil {
			return fmt.Sprintf("script:%s:%s", pathMatch[1], strings.TrimSpace(patternMatch[1]))
		}
		if pathMatch != nil {
			return fmt.Sprintf("script:%s", pathMatch[1])
		}
		return fmt.Sprintf("script:%s", stripped)
	}

	if section == "URL Rewrite" {
		parts := strings.SplitN(stripped, ",", 2)
		return fmt.Sprintf("rewrite:%s", parts[0])
	}
	if section == "Map Local" {
		return fmt.Sprintf("maplocal:%s", stripped)
	}
	if section == "Body Rewrite" {
		return fmt.Sprintf("body:%s", stripped)
	}
	if section == "Header Rewrite" {
		return fmt.Sprintf("header:%s", stripped)
	}
	if section == "Rule" {
		return fmt.Sprintf("rule:%s", stripped)
	}
	if section == "MITM" {
		re := regexp.MustCompile(`(?i)^hostname\s*=\s*(%APPEND%\s*)?`)
		hosts := re.ReplaceAllString(stripped, "")
		return fmt.Sprintf("mitm:%s", strings.TrimSpace(hosts))
	}
	return fmt.Sprintf("%s:%s", section, stripped)
}

type parsedLine struct {
	raw   string
	left  string
	right string
}

func DedupeSectionLines(section string, lines []string) []string {
	seen := make(map[string]bool)
	var result []string
	for _, line := range lines {
		stripped := strings.TrimSpace(line)
		if stripped == "" {
			if len(result) > 0 && result[len(result)-1] != "" {
				result = append(result, "")
			}
			continue
		}
		if strings.HasPrefix(stripped, "#") {
			result = append(result, stripped)
			continue
		}
		key := dedupeKey(section, line)
		if key == "" || seen[key] {
			continue
		}
		seen[key] = true
		result = append(result, stripped)
	}

	if section == "Script" || section == "Host" || section == "MITM" || section == "General" {
		var parsed []parsedLine
		maxLen := 0
		for _, r := range result {
			if strings.HasPrefix(r, "#") {
				parsed = append(parsed, parsedLine{raw: r})
				continue
			}
			parts := strings.SplitN(r, "=", 2)
			if len(parts) == 2 {
				left := strings.TrimSpace(parts[0])
				right := strings.TrimSpace(parts[1])
				if len(left) < 60 {
					if len(left) > maxLen {
						maxLen = len(left)
					}
				}
				parsed = append(parsed, parsedLine{raw: r, left: left, right: right})
			} else {
				parsed = append(parsed, parsedLine{raw: r})
			}
		}

		var alignedResult []string
		for _, p := range parsed {
			if p.left != "" && p.right != "" && len(p.left) <= 60 {
				alignedResult = append(alignedResult, fmt.Sprintf("%-*s = %s", maxLen, p.left, p.right))
			} else {
				alignedResult = append(alignedResult, p.raw)
			}
		}
		return alignedResult
	}

	return result
}

func FormatHeader(meta map[string]string, extraLines []string) []string {
	var lines []string
	used := make(map[string]bool)
	for _, key := range HeaderKeysOrder {
		if val, ok := meta[key]; ok {
			lines = append(lines, fmt.Sprintf("#!%s=%s", key, val))
			used[key] = true
		}
	}

	// Sorted remaining keys
	var remaining []string
	for key := range meta {
		if !used[key] {
			remaining = append(remaining, key)
		}
	}
	sort.Strings(remaining)
	for _, key := range remaining {
		lines = append(lines, fmt.Sprintf("#!%s=%s", key, meta[key]))
	}

	if len(extraLines) > 0 {
		lines = append(lines, extraLines...)
	}
	return lines
}

func FormatModule(headerLines []string, sections []ModuleSection, dedupe bool, sectionComments map[string]string) string {
	sectionMap := make(map[string][]string)
	for _, sec := range sections {
		sectionMap[sec.Name] = sec.Lines
	}

	var out []string
	out = append(out, headerLines...)
	for len(out) > 0 && strings.TrimSpace(out[len(out)-1]) == "" {
		out = out[:len(out)-1]
	}
	if len(out) > 0 {
		out = append(out, "")
	}

	for _, sectionName := range SectionOrder {
		lines, ok := sectionMap[sectionName]
		if !ok || len(lines) == 0 {
			continue
		}
		if dedupe {
			lines = DedupeSectionLines(sectionName, lines)
		}
		if len(lines) == 0 {
			continue
		}
		if comment, ok := sectionComments[sectionName]; ok {
			out = append(out, comment)
		}
		out = append(out, fmt.Sprintf("[%s]", sectionName))
		out = append(out, lines...)
		out = append(out, "")
	}

	res := strings.Join(out, "\n")
	res = strings.TrimRight(res, "\n") + "\n"
	return res
}

func SanitizeFileContent(text string, dedupe bool) string {
	meta, sections := ParseModule(text)
	var header []string
	if len(meta) > 0 {
		header = FormatHeader(meta, nil)
	} else {
		lines := strings.Split(text, "\n")
		for _, raw := range lines {
			if strings.HasPrefix(strings.TrimSpace(raw), "[") {
				break
			}
			if strings.HasPrefix(strings.TrimSpace(raw), "#!") || strings.TrimSpace(raw) == "" {
				header = append(header, strings.TrimRight(raw, "\r"))
			}
		}
	}
	return FormatModule(header, sections, dedupe, nil)
}

func MergeMitmHosts(sections []ModuleSection) []ModuleSection {
	hosts := make(map[string]bool)
	var other []ModuleSection
	skipTokens := map[string]bool{"%INSERT%": true, "%APPEND%": true}

	for _, sec := range sections {
		if sec.Name != "MITM" {
			other = append(other, sec)
			continue
		}
		for _, line := range sec.Lines {
			stripped := strings.TrimSpace(line)
			if strings.HasPrefix(strings.ToLower(stripped), "hostname") {
				re1 := regexp.MustCompile(`(?i)^hostname\s*=\s*`)
				part := re1.ReplaceAllString(stripped, "")
				re2 := regexp.MustCompile(`(?i)^(%APPEND%|%INSERT%)\s*`)
				part = re2.ReplaceAllString(part, "")
				for _, token := range strings.Split(part, ",") {
					token = strings.TrimSpace(token)
					if token != "" && !skipTokens[token] {
						hosts[token] = true
					}
				}
			}
		}
	}

	if len(hosts) > 0 {
		var inclusions, exclusions []string
		for h := range hosts {
			if strings.HasPrefix(h, "-") {
				exclusions = append(exclusions, h)
			} else {
				inclusions = append(inclusions, h)
			}
		}
		sort.Strings(inclusions)
		sort.Strings(exclusions)
		merged := append(inclusions, exclusions...)

		other = append(other, ModuleSection{
			Name:  "MITM",
			Lines: []string{fmt.Sprintf("hostname = %%APPEND%% %s", strings.Join(merged, ", "))},
		})
	}
	return other
}

File: create/scripts/hub/module_sanitizer.py
#!/usr/bin/env python3
"""
Normalize Surge / Shadowrocket module files: dedupe sections, preserve #!arguments,
standard section order, and readable grouping comments.
"""

from __future__ import annotations

import re
from typing import Dict, List, Optional, Set, Tuple

SECTION_ORDER = [
    "General",
    "Host",
    "Rule",
    "URL Rewrite",
    "Map Local",
    "Script",
    "Body Rewrite",
    "Header Rewrite",
    "MITM",
]

HEADER_KEYS_ORDER = [
    "name",
    "desc",
    "author",
    "icon",
    "category",
    "tag",
    "date",
    "arguments",
    "arguments-desc",
    "system-proxy",
    "ability",
    "update-interval",
]

META_LINE_RE = re.compile(r"^#!\s*([A-Za-z0-9_-]+)\s*[=:]\s*(.*)$", re.IGNORECASE)
SCRIPT_NAME_RE = re.compile(r"\bname\s*=\s*([^,\s]+)", re.IGNORECASE)
SCRIPT_LABEL_RE = re.compile(r"^(.+?)\s*=\s*type=", re.IGNORECASE)
SCRIPT_PATH_RE = re.compile(r"\bscript-path\s*=\s*([^,\s]+)", re.IGNORECASE)
SCRIPT_PATTERN_RE = re.compile(r"\bpattern\s*=\s*([^,]+)", re.IGNORECASE)


def parse_module(text: str) -> Tuple[Dict[str, str], List[Tuple[str, List[str]]]]:
    """Return (header_meta, sections) where header_meta keys are lowercased."""
    header_meta: Dict[str, str] = {}
    sections: List[Tuple[str, List[str]]] = []
    current_name: Optional[str] = None
    current_lines: List[str] = []
    in_body = False

    for raw in text.splitlines():
        stripped = raw.strip()
        if stripped.startswith("[") and stripped.endswith("]") and not stripped.startswith("#"):
            if current_name is not None:
                sections.append((current_name, current_lines))
            current_name = stripped[1:-1]
            current_lines = []
            in_body = True
            continue

        if not in_body:
            match = META_LINE_RE.match(stripped)
            if match:
                header_meta[match.group(1).lower()] = match.group(2).strip()
            elif stripped.startswith("#!") or not stripped:
                pass
            continue

        if current_name is not None:
            current_lines.append(raw.rstrip())

    if current_name is not None:
        sections.append((current_name, current_lines))
    return header_meta, sections


def _dedupe_key(section: str, line: str) -> str:
    stripped = line.strip()
    if not stripped or stripped.startswith("#"):
        return ""
    lowered = stripped.lower()

    if section == "Script":
        label = SCRIPT_LABEL_RE.match(stripped)
        if label:
            return f"script:label:{label.group(1).strip()}"
        name = SCRIPT_NAME_RE.search(stripped)
        if name:
            return f"script:name:{name.group(1)}"
        path = SCRIPT_PATH_RE.search(stripped)
        pattern = SCRIPT_PATTERN_RE.search(stripped)
        if path and pattern:
            return f"script:{path.group(1)}:{pattern.group(1).strip()}"
        if path:
            return f"script:{path.group(1)}"
        return f"script:{stripped}"

    if section == "URL Rewrite":
        return f"rewrite:{stripped.split(',')[0] if ',' in stripped else stripped}"

    if section == "Map Local":
        return f"maplocal:{stripped}"

    if section == "Body Rewrite":
        return f"body:{stripped}"

    if section == "Header Rewrite":
        return f"header:{stripped}"

    if section == "Rule":
        return f"rule:{stripped}"

    if section == "MITM":
        hosts = re.sub(r"^hostname\s*=\s*(%APPEND%\s*)?", "", stripped, flags=re.I)
        return f"mitm:{hosts.strip()}"
    return f"{section}:{stripped}"

def dedupe_section_lines(section: str, lines: List[str]) -> List[str]:
    seen: Set[str] = set()
    result: List[str] = []
    for line in lines:
        stripped = line.strip()
        if not stripped:
            if result and result[-1] != "":
                result.append("")
            continue
        if stripped.startswith("#"):
            result.append(stripped)
            continue
        key = _dedupe_key(section, line)
        if not key or key in seen:
            continue
        seen.add(key)
        result.append(stripped)

    # Beautify alignment for sections with key=value formats
    if section in ("Script", "Host", "MITM", "General"):
        parsed = []
        max_len = 0
        for r in result:
            if r.startswith("#"):
                parsed.append((r, None, None))
                continue
            parts = r.split("=", 1)
            if len(parts) == 2:
                left = parts[0].strip()
                right = parts[1].strip()
                # To prevent absurd padding from malformed lines, cap max_len at 60
                if len(left) < 60:
                    max_len = max(max_len, len(left))
                parsed.append((r, left, right))
            else:
                parsed.append((r, None, None))
        
        aligned_result = []
        for r, left, right in parsed:
            if left is not None and right is not None and len(left) <= 60:
                aligned_result.append(f"{left:<{max_len}} = {right}")
            else:
                aligned_result.append(r)
        return aligned_result

    return result


def format_header(meta: Dict[str, str], extra_lines: Optional[List[str]] = None) -> List[str]:
    lines: List[str] = []
    used = set()
    for key in HEADER_KEYS_ORDER:
        if key in meta:
            lines.append(f"#!{key}={meta[key]}")
            used.add(key)
    for key, value in sorted(meta.items()):
        if key not in used:
            lines.append(f"#!{key}={value}")
    if extra_lines:
        lines.extend(extra_lines)
    return lines


def format_module(
    header_lines: List[str],
    sections: List[Tuple[str, List[str]]],
    *,
    dedupe: bool = True,
    section_comments: Optional[Dict[str, str]] = None,
) -> str:
    """Build module text with standard section order."""
    section_map = {name: lines for name, lines in sections}
    out: List[str] = list(header_lines)
    while out and not out[-1].strip():
        out.pop()
    if out:
        out.append("")

    for section_name in SECTION_ORDER:
        lines = section_map.get(section_name)
        if not lines:
            continue
        if dedupe:
            lines = dedupe_section_lines(section_name, lines)
        if not lines:
            continue
        if section_comments and section_name in section_comments:
            out.append(section_comments[section_name])
        out.append(f"[{section_name}]")
        out.extend(lines)
        out.append("")

    return "\n".join(out).rstrip() + "\n"


def sanitize_file_content(text: str, *, dedupe: bool = True) -> str:
    meta, sections = parse_module(text)
    header = format_header(meta) if meta else []
    if not header:
        # Legacy: keep non-meta preamble lines before first section
        header = []
        for raw in text.splitlines():
            if raw.strip().startswith("["):
                break
            if raw.strip().startswith("#!") or not raw.strip():
                header.append(raw.rstrip())
    return format_module(header, sections, dedupe=dedupe)


def merge_mitm_hosts(sections: List[Tuple[str, List[str]]]) -> List[Tuple[str, List[str]]]:
    hosts: Set[str] = set()
    other: List[Tuple[str, List[str]]] = []
    skip_tokens = {"%INSERT%", "%APPEND%"}
    for name, lines in sections:
        if name != "MITM":
            other.append((name, lines))
            continue
        for line in lines:
            stripped = line.strip()
            if stripped.lower().startswith("hostname"):
                part = re.sub(r"^hostname\s*=\s*", "", stripped, flags=re.I)
                part = re.sub(r"^(%APPEND%|%INSERT%)\s*", "", part, flags=re.I)
                for token in part.split(","):
                    token = token.strip()
                    if token and token not in skip_tokens:
                        hosts.update({token})
    if hosts:
        inclusions = sorted(h for h in hosts if not h.startswith("-"))
        exclusions = sorted(h for h in hosts if h.startswith("-"))
        merged = ", ".join(inclusions + exclusions)
        other.append(("MITM", [f"hostname = %APPEND% {merged}"]))
    return other
